### PR TITLE
Optimized NumPy array as keys

### PR DESCRIPTION
**Pull request by flexatone**

### CODE REVIEW

**Comment by flexatone** [comment]
FYI: this is very preliminary work... I meant to push to my fork!

**Comment by flexatone** [comment]
Initial performance panel before any material changes.
automap

**Comment by flexatone** [comment]
This iteration holds on to an immutable 1D array as the internal key store.
This generally results in initialization faster than calling "tolist()" on an array, but not faster than just giving a list. For most types (float, int, string) lookup and items iteration is much slower when the internal keys are a list, as are converting NumPy elements to PyObjects on every iteration.

This suggest that storing an array as keys is not the best approach. Next, I will look automatically calling "tolist" on arrays, and possibly exposing an "own_keys" argument.

automap

**Comment by flexatone** [comment]
Replacing all usage of `PyArray_GETITEM` with `PyArray_ToScalar` (necessary to preserve `datetime64` types seems to degrade performance, which is consistent with observations that `a.tolist()` is faster than `list(a)`.

automap

**Comment by flexatone** [comment]
When an `AutoMap`  gets an array, regardless if it is immutable, we use either  `PySequence_List` (if a dt64 type) or `PyArray_ToList` and store the keys as a list. This results in initialization performance equivalent to initialization from an arbitrary list.
automap

**Comment by flexatone** [comment]
I have succeeded in optimizing handling of internally stored integer arrays, which now show extremely fast initialization and as fast-as-dict lookup; items iterations is still slower but this is likely acceptable (and improvable by using `PyArray_GETITEM`.

The opportunity here is using the integer as its hash, and as such avoiding computing hashes, comparing Python objects, and creating objects for each key.
automap

**Comment by flexatone** [comment]
Added support for variable sized integers.
automap

**Comment by flexatone** [comment]
Implemented usage of NumPy float arrays for direct hashing and lookup. We see modest better-than list performance for instantiation, and seemingly better lookup performance.
automap

**Comment by flexatone** [comment]
After implementing support for NumPy Unicode strings and a custom hashing method for raw Py_UCS4 arrays, initialization is far faster than before and now faster than initialization from lists. Keys iterations suffers, but that tradeoff is acceptable.
automap

**Comment by flexatone** [comment]
After implementing a re-usable Py-UCS4 buffer that lives on `fam` and is used to store key data on lookup seems to provide a modest benefit. The benefit is likely to be greater with bigger strings, so it seems worth keeping.
automap

**Comment by flexatone** [comment]
implemented new loading insertion routines that take advantage of contiguous arrays when possible. Seems to deliver some benefit for large inputs.

New performance measures show poor performance of "not in" checks on floats and ints; something to look at. 

automap

**Comment by flexatone** [comment]
Extended performance tests for bytes and uints. Removed outlier not-in performance on numeric array types by using `PyNumber_Check` before calling `PyNumber_AsSsize_t`.
automap

**Comment by flexatone** [comment]
Using `memcmp` in Unicode lookup might provide some benefit. 
automap
Fixed order of types in display.

**Comment by flexatone** [comment]
Implemented support for immutable bytes arrays. Massive performance improvement on instantiation and lookup.

automap

**Comment by flexatone** [comment]
This performance test shows no noticeable change due to the addition of checks for Scalars for integers and floats.

automap

**Comment by flexatone** [comment]
@brandtbucher : as you suggested, the performance advantage FAM(L) had in lookup routines was only due to doing those lookups with the same objects stored in the FAM's keys, where an object id check can give excellent performance.

If we do the lookups with NumPy array scalar ("lookup AS" in the charts below), FAM(A) is as fast or faster for all types.

automap

**Comment by flexatone** [comment]
Current performance benchmarks, now split into two displays. Note that the relevant comparison is between FAM(A) and AM(A), the latter approximating what StaticFrame is doing presently. FAM(L) is not available to StaticFrame as AutoMaps are always created from an array, not a list of Python objects. In all cases, FAM(A) now out-performs AM(A) for both initialization and lookup.

automap-instantiate
automap-lookup

**Comment by flexatone** [comment]
## Overview

These changes integrate direct support of NumPy arrays given as keys to `AutoMap`s and `FrozenAutoMap`s, optimizing their usage.

Improvements are made in `AutoMap` initialization, whereby an array is converted to a list using optimal array methods; that list is then held as the keys.

Improvements are made in `FrozenAutoMap` initialization, whereby an immutable array (for integer, floating point, and flexible dtypes), when given as keys, is held as a reference without copying to a list. Further, hashing and lookup make use of C types, avoiding any management of PyObjects.

For array dtypes not explicitly handled, or for non-array keys, `FrozenAutoMap` operation is unchanged. In all cases, hash table layout and scanning, and management of the PyObject integer cache, are the same as before.

A key change is that, on initialization (`fam_init`), a `KeysArrayType` enum value is assigned to the `keys_array_type` attribute of `FAMObject`. This is used for branching in all places where divergent behavior is required between keys stored as lists (as was done previously) or as keys stored as typed arrays.

Performance panels (https://github.com/brandtbucher/automap/pull/16#issuecomment-1487601017) compare FAM(L), FAM(A), AM(A), and Dict (`FrozenAutoMap` created from a list, `FrozenAutoMap` created from an array, `AutoMap` created from an array, and a dictionary implementing an `AutoMap` mapping). Key indicators are the performance of instantiation and lookup.

The relevant comparison for StaticFrame usage is between FAM(A) and AM(A), the latter approximating what StaticFrame does presently when creating `FrozenAutoMap`s. (FAM(L) is not available to StaticFrame as AutoMaps are always created from an array, not a list of Python objects.)

Across all supported types, FAM(A) initialization is more than twice as fast as AM(A). FAM(A) lookup performance varies greatly by type, but always out-performs AM(A), in some cases more than twice as fast as AM(A). In all tests, we see signs that out-performance grows with scale.

Independent of performance time, All `FrozenAutoMap` usage of arrays reduces memory usage: no new `PyObject`s are created and the passed array simply has reference incremented.

## Key Changes

Split the old `fam_new()` into `fam_new()` and `fam_init()`, implemented `__setstate__()`, `__getstate__()`:
* To support pickling a FAM with a NumPy array, `__setstate__()` must re-set the `writeable` flag of an arary to False.
* To integrate `__setstate__()`, the old `fam_new()` had to be divided into a `fam_new()` and a `fam_init()`.

Based on `keys_array_type`, `fam_init` calls type-specific insert routines, which use type-specific hash routines to add entries to the hash table.

On lookup, type-specific lookup routines are called based on `keys_array_type`. These routines identify PyObjects as PyArray scalars or native PyObject types, extract the appropriate C-type, compute a hash, and use type-specific lookup routines to discover the position in the keys array.


Split `copy()` into `copy()` and `copy_to_new()`.
* Due to now having `fam_new()` and `fam_init()`, copy allocation and copy setting needed to split into two methods. Now, in `fam_init`, if a `FAMType` is identified as the keys argument, `copy_to_new()` is used to transfer values from the argument to the new instance. The `copy()` function remains, now using `fam_new()` and `copy_to_new()`.

Additions to the `FAMObject` struct:
* `key_array_type`: A `KeysArrayType` enum specifying a list or array dtype. As a list is assigned zero (and all other array dtypes as non-zero), the value can be used to branch on non-array versus array processing.
* `keys_size`: As determining size must branch on `keys_array_type`, this attribute is used to track size, avoiding having to go the underly keys container.
* `key_buffer`: For Unicode arrays, this is a dynamically allocated buffer of size equal to one more than the array's max number of characters. This buffer is given to `PyUnicode_AsUCS4`, which will add a NULL and is why the size of the buffer is one more than max characters. This is only used for a FAM with Unicode array keys; all other usage keeps this as NULL.

The type of `key_count_global` is now a platform independent 64 bit signed integer. Perviously, it was a `Py_ssize_t`, which is 32 bits on 32 bit systems and could overflow in scenarios when many indicies are created.

Extended property tests for FAMs with arrays
* A custom Hypothesis strategy has been implemented to deliver both contiguous and non-contiguous arrays.
* New Hypothesis tests for array-initialized `FrozenAutoMap`s now cover all features previously tested by Hypothesis.

## Next Steps

I welcome all feedback. Thanks to @brandtbucher for the core implementation and preliminary comments on this extension. Thanks also to @chaburkland for helping me work through a number of issues with this code.

I recognize that this is a substantial expansion of the code for this package, and I am happy to contribute to ongoing maintenance to support these changes. That said, I understand that these changes might exceed the goals of this package;  if so, I can certainly explore alternative distribution and packaging (though I would still be grateful for any feedback!). 

After these changes are integrated into StaticFrame, I would like to explore direct support for `datetime64` arrays in `FrozenAutoMap`s, where performance can similarly be greatly improved.

File: automap.c [hunk 1]
```diff
@@ -832,78 +1814,231 @@ fam_get(FAMObject *self, PyObject *args)
...
+            return -1;
+        }
+        int array_t = PyArray_TYPE(a);
+        if (cls != &AMType &&
+                (PyTypeNum_ISINTEGER(array_t) // signed and unsigned
+                || PyTypeNum_ISFLOAT(array_t)
+                || PyTypeNum_ISFLEXIBLE(array_t))
+            ){
+            if ((PyArray_FLAGS(a) & NPY_ARRAY_WRITEABLE)) {
+                PyErr_Format(PyExc_TypeError, "Arrays must be immutable when given to a %s", name);
```
Comment (chaburkland) [inline automap.c:1944]: This is not what I would expect the behavior to be. I would imagine it simply checks if the array is immutable. If it is not, then just proceed using the old code.

I think it's unexpected (and undesirable) behavior to fail instantiating a `FrozenAutoMap` when given a numpy array that is not immutable. This seems like the responsibility of the client, not automap.

File: automap.c [hunk 2]
```diff
@@ -832,78 +1814,231 @@ fam_get(FAMObject *self, PyObject *args)
...
-    {NULL},
-};
-
+// This macro can be used with integer and floating point NumPy types, given an `npy_type` and a specialized `insert_func`. Uses context of `fam_init` to get `fam`, `contiguous`, `a`, `keys_size`, and `i`. An optional `pre_insert` function can be supplied to transform extracted values before calling the insert function.
+# define INSERT_SCALARS(npy_type, insert_func, pre_insert)    \
+if (contiguous) {                                             \
+    npy_type* b = (npy_type*)PyArray_DATA(a);                 \
+    npy_type* b_end = b + keys_size;                          \
+    while (b < b_end) {                                       \
+        if (insert_func(fam, pre_insert(*b), i, -1)) {        \
```
Comment (chaburkland) [inline automap.c:1841]: `pre_insert(*b)` -> `(*b)` when `pre_insert` is empty is really clever

File: automap.c [hunk 3]
```diff
@@ -110,74 +108,304 @@ objects from value lookups), but the hardware-friendly hash table design is what
...
+static npy_int64 key_count_global = 0;
+
+
+// Fill the int_cache up to size_needed with PyObject ints; `size` is not the key_count_global.
+static int
+int_cache_fill(Py_ssize_t size_needed)
+{
+    PyObject *item;
+    if (!int_cache) {
+        int_cache = PyList_New(0);
```
[the remark this answers is not in the document]
Comment (brandtbucher) [reply automap.c:358]: The main reason is because that would require two loops: one using `PyList_SET_ITEM` for the uncommon case where the list is newly-created, and one using `PyList_Append` for all other resizes. I didn't give it too much thought, and just went with what was simplest.

[the remark this answers is not in the document]
Comment (brandtbucher) [reply automap.c:358]: (It looks like this change was copied from the original, with some renamings.)

File: automap.c [hunk 4]
```diff
@@ -386,256 +642,968 @@ famv_isdisjoint(FAMVObject *self, PyObject *other)
...
+                continue;
+            }
+            p_start = (char*)PyArray_GETPTR1(a, table[table_pos].keys_pos);
+            // memcmp returns 0 on match
+            if (!memcmp(p_start, key, (key_size < dt_size ? key_size : dt_size))) {
+                return table_pos;
+            }
+            table_pos++;
+        }
+        table_pos = (5 * (table_pos - SCAN) + (mixin >>= 1) + 1) & mask;
```
[the remark this answers is not in the document]
Comment (brandtbucher) [reply automap.c:962]: This is correct. As it's shifted down, `mixin` will start including higher bits of the hash that were ignored during the original index calculation. It's a way of introducing entropy for hashes that collide in the low bits but not the higher ones.

If it truly gets shifted down to zero (if we get *very* unlucky with our table layout and hit over 1,000 colliding items), it means that we have no more entropy in the hash left to use.  If that's the case, then using zeroes for the `mixin` value is (basically) just as useful as using any other bit pattern.

**More importantly, though, it's also required for correctness that `mixin` eventually becomes zero!** The sequence `i = (5 * i + some_random_number + 1) % tablesize` is *not* guaranteed to visit every spot in a table with a size that's a power of two (and it may even visit some spots many times). However, `i = (5 * i + 1) % tablesize` is (and each spot will be visited *exactly* once)! By ensuring that `mixin` eventually becomes (and stays) zero, we can guarantee that all lookups terminate eventually (in the worst case, by visiting every single occupied slot in the table before finally hitting an empty one).

[the remark this answers is not in the document]
Comment (brandtbucher) [reply automap.c:962]: [Further reading.](https://github.com/python/cpython/blob/main/Objects/dictobject.c#L165-L217)

[the remark this answers is not in the document]
Comment (chaburkland) [reply automap.c:962]: That makes a lot of sense - thanks! :)

File: automap.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (brandtbucher) [reply automap.c:358]: It’s almost certainly faster, but it only happens literally once per process. So I just felt that it wasn’t really worth adding a whole second code path, since there’s no way that it’s a bottleneck (and benchmarking the tiny improvement is super difficult).

File: automap.c [hunk 5]
```diff
@@ -386,256 +642,968 @@ famv_isdisjoint(FAMVObject *self, PyObject *other)
...
+    Py_INCREF(fam);
+    return (PyObject *)famv;
+}
+
+//------------------------------------------------------------------------------
+// FrozenAutoMap functions
+
+// Given a key and a computed hash, return the table_pos if that hash and key are found, or if not, the first table position that has not been assigned. Return -1 on error.
+static Py_ssize_t
+lookup_hash(FAMObject *self, PyObject *key, Py_hash_t hash)
```
[the remark this answers is not in the document]
Comment (flexatone) [reply automap.c:703]: Yes, that does seem more clear; it might also suggest some other re-namings.

File: automap.c [hunk 6]
```diff
@@ -190,37 +418,52 @@ fami_iter(FAMIObject *self)
...
+                    PyList_GET_ITEM(int_cache, index)
+                );
+            }
         }
         case KEYS: {
-            PyObject *yield = PyList_GET_ITEM(self->map->keys, index);
-            Py_INCREF(yield);
-            return yield;
+            if (self->fam->keys_array_type) {
+                return PyArray_ToScalar(PyArray_GETPTR1(self->keys_array, index), self->keys_array);
```
[the remark this answers is not in the document]
Comment (flexatone) [reply automap.c:457]: Good question. My expectation is that this is a new reference as these scalar objects do not exist until a method like this is called: they have to be newly created references. What about the source suggests to you that they are not new references?

File: automap.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (flexatone) [reply automap.c:962]: Thanks, @brandtbucher , for explaining this!

File: automap.c [hunk 7]
```diff
@@ -386,256 +642,968 @@ famv_isdisjoint(FAMVObject *self, PyObject *other)
...
+            }
+            table_pos++;
+        }
+        table_pos = (5 * (table_pos - SCAN) + (mixin >>= 1) + 1) & mask;
+    }
+}
+
+
+static Py_ssize_t
+lookup_int(FAMObject *self, PyObject* key) {
```
[the remark this answers is not in the document]
Comment (flexatone) [reply automap.c:968]: Thanks for this suggestion. I originally had something like that, but I realized it was optimizing for some very unlikely cases: we are far more likely to get a lookup from a PyLong or PyFloat or a Double scalar than a ULongLong, Int, or UInt. So I tried to but the most likely types at  top at the cost of making the branches less coherent. This seemed to improve float lookup performance but I would have to do a discrete A/B test to confirm.

File: automap.c [hunk 8]
```diff
@@ -386,256 +642,968 @@ famv_isdisjoint(FAMVObject *self, PyObject *other)
...
+            if (h == -1) {
+                return table_pos;
+            }
+            if (h != hash) {
+                table_pos++;
+                continue;
+            }
+            p_start = (char*)PyArray_GETPTR1(a, table[table_pos].keys_pos);
+            // memcmp returns 0 on match
+            if (!memcmp(p_start, key, (key_size < dt_size ? key_size : dt_size))) {
```
[the remark this answers is not in the document]
Comment (flexatone) [reply automap.c:957]: My understanding is that there is not a standard `min` function in C99: is there?

File: automap.c [hunk 9]
```diff
@@ -386,256 +642,968 @@ famv_isdisjoint(FAMVObject *self, PyObject *other)
...
+            if (h == -1) {
+                return table_pos;
+            }
+            if (h != hash) {
+                table_pos++;
+                continue;
+            }
+            p_start = (Py_UCS4*)PyArray_GETPTR1(a, table[table_pos].keys_pos);
+            // memcmp returns 0 on match
+            if (!memcmp(p_start, key, (key_size < dt_size ? key_size : dt_size))) {
```
[the remark this answers is not in the document]
Comment (flexatone) [reply automap.c:916]: Same as above.

File: automap.c [hunk 10]
```diff
@@ -832,78 +1814,231 @@ fam_get(FAMObject *self, PyObject *args)
...
-    {"__sizeof__", (PyCFunction) fam___sizeof__, METH_NOARGS, NULL},
-    {"get", (PyCFunction) fam_get, METH_VARARGS, NULL},
-    {"items", (PyCFunction) fam_items, METH_NOARGS, NULL},
-    {"keys", (PyCFunction) fam_keys, METH_NOARGS, NULL},
-    {"values", (PyCFunction) fam_values, METH_NOARGS, NULL},
-    {NULL},
-};
-
+// This macro can be used with integer and floating point NumPy types, given an `npy_type` and a specialized `insert_func`. Uses context of `fam_init` to get `fam`, `contiguous`, `a`, `keys_size`, and `i`. An optional `pre_insert` function can be supplied to transform extracted values before calling the insert function.
+# define INSERT_SCALARS(npy_type, insert_func, pre_insert)    \
```
[the remark this answers is not in the document]
Comment (flexatone) [reply automap.c:1836]: Sure, good point. This is an artifact of when there was only a `fam_new`.

File: automap.c (hunk 1 above)
Comment (flexatone) [reply automap.c:1944]: Are you sure? As the performance characteristics are far different between using immutable arrays and converting those arrays to lists, I think it is important to only support one way of creating a `FAM` with supported array types. Otherwise, one could silently fall into using the old code simply because the client forgot to set the array as immutable.

File: automap.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (brandtbucher) [reply automap.c:957]: Use this: https://docs.python.org/3/c-api/intro.html#c.Py_MIN.

It's just a macro for this pattern, but any decent compiler will recognize the pattern and optimize it.

File: automap.c (hunk 2 above)
Comment (flexatone) [reply automap.c:1841]: Thanks! I was worried that the macro usage here would fall apart with the one-off need for converter function, but being able to use empty macro arguments is pretty nice!

File: automap.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (flexatone) [reply automap.c:1836]: This is now corrected, and I also `undef` those macros after usage.

File: automap.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (flexatone) [reply automap.c]: While adding `lookup_hash_obj` I also renamed `insert` to `insert_obj` for the same reasons.

File: automap.c (hunk 1 above)
Comment (chaburkland) [reply automap.c:1945]: I do feel rather strongly about it, as my expectations for `automap` is that it's designed to be a drop-in replacement for sets.

In other words:
`AutoMap` <=> `set`
`FrozenAutoMap` <=> `frozenset`

So, just like you can do `frozenset(mutable_numpy_arr)`, I would expect you could also do `FrozenAutoMap(mutable_numpy_arr)`

File: automap.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (flexatone) [reply automap.c]: `Py_MIN`!

File: automap.c (hunk 1 above)
Comment (flexatone) [reply automap.c:1945]: But why would it benefit the caller to support sub-optimal usage from a mutable array? The caller probably does not want that, and would rather get an error telling them to make the array immutable. If we really wanted to provide this convenience, we could make a copy of the array ourselves and set it to be immutable, but that would also introduce a silent performance penalty (but would probably still deliver better performance than converting the array to PyObjects).

Comment (chaburkland) [reply automap.c:1945]: I guess it's my opinion that it's not `automap`'s call to make. This is a great optimization, but it seems like an anti-pattern to force callers to modify their input in order to take advantage of this optimization

I think it would be better to have this be a `PerformanceWarning`, that lets users know about the overhead of using mutable arrays.

I feel strongly that it's the client's responsibility to opt-in, rather than automap's responsibility to disallow mutable arrays entirely

Comment (flexatone) [reply automap.c:1945]: Issuing a warning at least makes the undesirable usage not silent! I see your point that it is unconventional for the container to reject such input. But I also do not see `FAM` as having much general-purpose usage outside of StaticFrame... in the context of StaticFrame, I want a failure if I accidentally give a `FAM` a mutable array!

**Review by brandtbucher — COMMENTED** [review]
Thanks for your patience with the review process! I appreciate the amount of time and energy that goes into a project like this.

So, first things first... I started reviewing this code last night, but I quickly needed to stop because I was finding the diff *way* too hard to review. The main reason seems to be that this branch contains lots of unrelated reorganizing, renaming, etc. In an already huge PR, this sort of stuff makes the changes *really* hard to follow. As just one example, the diffs for the old `lookup` and the new `insert_double` are tangled together in GitHub's review UI. Lots of structure definitions, etc. are also moved around, making it very difficult to see what changed. Even little renamings like "intcache" to "int_cache" and changed whitespace are added lots of friction to the review process.

Is there any way that you could clean up or revert all but the absolutely necessary changes? Or something? I'm not sure what the best solution is, but I honestly don't think I'll be able to review this as-is. :(

Besides that, here are some more general thoughts I have when considering the design itself (as you've described it):

**1. This logic should all live in a subclass.** Perhaps the subclass could even be in a different module (or package). This also solves the issue of how "strict" the constructor should be when given a mutable array. The subclass can be strict, and the superclass can be lenient (as it always has). It also keeps this intimate knowledge of the NumPy internals out of the core data structure.

It may *seem* that this will just result in the subclass reimplementing all of the core logic from the base class, but that doesn't necessarily have to be true. Which brings me to my next point:

**2. To the greatest extent possible, the new NumPy-specific code should avoid touching the core data structure.** It's currently very easy to reason about, tweak, and fix bugs the core table implementation and experiment with new designs. The new code makes it much, much harder. To facilitate this, the superclass could be changed to be more friendly to subclasses... for example, what if all AutoMaps had two function pointers on them for retrieving items and computing hashes? Subclasses could set these functions to be whatever they want, without needing to re-implement the entire table logic, iterators, views, etc. This also opens up the possiblility of having dedicated subclasses for other immutable key arrays, like tuples or memoryviews.

**3. Lastly, but most importantly, I'm not yet convinced by the performance numbers.** After I pointed out that the disappointing lookup benchmarks were hitting the current code's fast path, the benchmarking code was updated to use the slowest possible path 100% of the time (resulting in the current numbers). It seems a bit disingenuous to benchmark the old code in this way; the entire reason that the fast-path exists is because I believe that it's a common enough case to optimize. The new benchmarks seem to make the opposite assumption (that it never happens at all).

What would convince me that this is indeed the common case, would be instrumenting Static Frame's current code with counters around each AutoMap access, tracking (at least):
- What percentage of lookups use Python objects that are present in the keys by identity.
- What percentage of lookups use equal-but-not-identical Python objects. 
- What percentage of lookups use NumPy scalars that are present in the keys by identity.
- What percentage of lookups use equal-but-not-identical NumPy scalars.

Then, this instrumented SF should be used for some *real, production process*.

My (maybe naive) assumption is that data has few sources and many uses. So, I would expect that the majority of lookups would be for the same Python object or NumPy scalar (depending, of course, on how the AutoMaps themselves are constructed). If that's the case, then it would seem that the current lookup fast-path is indeed faster than reimplementing hashing and equality for most NumPy scalars. Even if it's not the majority, there may still be enough of them to make the current fast path worth keeping. At any rate, the benchmarks we use to inform this change should use a realistic blend of lookups, rather than just 100% one way or the other.

Barring that, I'm afraid that my overall attitude towards this project hasn't changed much since [my comment eight months ago](https://github.com/brandtbucher/automap/issues/6#issuecomment-1226392927), and this might be a better fit for vendoring in ArrayKit.

File: automap.c [hunk 11]
```diff
@@ -110,74 +108,304 @@ objects from value lookups), but the hardware-friendly hash table design is what
...
+int_to_hash(npy_int64 v) {
+    Py_hash_t hash = (Py_hash_t)v;
+    if (hash == -1) {
+        return -2;
+    }
+    return hash;
+}
+
+
+// This is a adapted from https://github.com/python/cpython/blob/ba65a065cf07a7a9f53be61057a090f7311a5ad7/Python/pyhash.c#L92
```
[the remark this answers is not in the document]
Comment (flexatone) [reply automap.c:269]: Copied, with some MACRO values replace or renamed. Fundamentally it is the same.

File: automap.c [hunk 12]
```diff
@@ -110,74 +108,304 @@ objects from value lookups), but the hardware-friendly hash table design is what
...
+
+
+//------------------------------------------------------------------------------
+// the global int_cache is shared among all instances
+
+static PyObject *int_cache = NULL;
+
+
+// NOTE: this used to be a Py_ssize_t, which can be 32 bits on some machines and might easily overflow with a few very large indices. Using an explicit 64-bit int seems safer
+static npy_int64 key_count_global = 0;
```
[the remark this answers is not in the document]
Comment (flexatone) [reply automap.c:349]: Yeah, I had the same thought, but as those two things must be aliased to the same thing, and we use NumPy types elsewhere (but `*_t` nowhere) I figured it was better to use the NumPy types.

File: automap.c [hunk 13]
```diff
@@ -190,37 +418,52 @@ fami_iter(FAMIObject *self)
...
+                    PyList_GET_ITEM(int_cache, index)
+                );
+            }
+            else {
+                return PyTuple_Pack(
+                    2,
+                    PyList_GET_ITEM(self->fam->keys, index),
+                    PyList_GET_ITEM(int_cache, index)
+                );
+            }
```
[the remark this answers is not in the document]
Comment (flexatone) [reply automap.c:453]: That is a nice improvement!

File: automap.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (brandtbucher) [reply automap.c:349]: I trust the C standard more than NumPy. ;)

**Comment by flexatone** [comment]
Thanks, @brandtbucher , for your time and comments.

I know that this is hard to review as diffs, as there are many changes and the code is more than twice as large as it was. Just reviewing the code as it is more practical than the diffs as it stands. 

Regarding your specific points:

(1&2): A subclass-based implementation is attractive! The effort seemed daunting, however, and the amount of code duplication necessary seemed about equivalent to the branching approach that I took. If there was demand for other subclasses of immutable key arrays, this might also be attractive, but I do not think there is.

(3) I need to integrate this into StaticFrame to conduct real performance tests to be sure of anything. But I disagree that the current benchmarks are disingenuous: I expect that most lookups in StaticFrame are either with NumPy Scalars (which, to, my knowledge, are unique instances for each access) or for integers outside of the range of Python cached integers and thus new instances. Remember that in StaticFrame we are almost always starting with arrays when creating AutoMaps, not lists of Python objects. Thus the likelihood that we lookup with the same element instance is, in my expectation, very low (or close to not happening at all!). 

Elaborating on the issue with NumPy Scalars, notice how we get unique objects each time we iterate through the array. This is why I expect same-element instance lookup to be actually very rare.

```python
>>> a1
array([3, 4, 5])
>>> [id(a1[i]) for i in range(len(a1))]
[140249480531792, 140248763421840, 140248763421840]
>>> [id(a1[i]) for i in range(len(a1))]
[140248763421840, 140248763421680, 140248763421680]
```

Independent of lookup performance, I also assume that instantiation is as (or more) important than lookup, given that many indices are created and never used. The instantiation performance (comparing between FAM(A) and AM(A) is (for me) very compelling.

Given your points and closing statement, I would like to proceed by distributing this through a different channel such that I can fully build and test in StaticFrame on all platforms. Then, if the performance is good, and If you have time, interest, and bandwidth to layout the subclassing approach, I would be happy to re-implement this into AutoMap.

That said, I welcome any additional feedback or suggestions (now or later) on what we have here: I really appreciate it.
